api.go

```diff
@@ -798,61 +845,89 @@ func (res *resource) handleDelete(w http.ResponseWriter, r *http.Request, ps htt
 	return nil
 }
 
-func writeResult(w http.ResponseWriter, data []byte, status int) {
-	w.Header().Set("Content-Type", "application/vnd.api+json")
+func writeResult(w http.ResponseWriter, data []byte, status int, contentType string) {
+	w.Header().Set("Content-Type", contentType)
 	w.WriteHeader(status)
 	w.Write(data)
 }
 
-func respondWith(obj interface{}, info information, status int, w http.ResponseWriter) error {
-	data, err := jsonapi.MarshalToJSONWithURLs(obj, info)
+func respondWith(obj interface{}, info information, status int, w http.ResponseWriter, r *http.Request, marshalers map[string]ContentMarshaler) error {
+	data, err := jsonapi.MarshalWithURLs(obj, info)
 	if err != nil {
 		return err
 	}
 
-	writeResult(w, data, status)
-	return nil
+	return marshalResponse(data, w, status, r, marshalers)
 }
 
-func respondWithPagination(obj interface{}, info information, status int, links map[string]string, w http.ResponseWriter) error {
+func respondWithPagination(obj interface{}, info information, status int, links map[string]string, w http.ResponseWriter, r *http.Request, marshalers map[string]ContentMarshaler) error {
 	data, err := jsonapi.MarshalWithURLs(obj, info)
 	if err != nil {
 		return err
 	}
 
 	data["links"] = links
-	result, err := json.Marshal(data)
-	if err != nil {
-		return err
-	}
-
-	writeResult(w, result, status)
-	return nil
+	return marshalResponse(data, w, status, r, marshalers)
 }
 
-func unmarshalJSONRequest(r *http.Request) (map[string]interface{}, error) {
+func unmarshalRequest(r *http.Request, marshalers map[string]ContentMarshaler) (map[string]interface{}, error) {
 	defer r.Body.Close()
 	data, err := ioutil.ReadAll(r.Body)
 	if err != nil {
 		return nil, err
 	}
 	result := map[string]interface{}{}
-	err = json.Unmarshal(data, &result)
+	marshaler, _ := selectContentMarshaler(r, marshalers)
+	err = marshaler.Unmarshal(data, &result)
 	if err != nil {
 		return nil, err
 	}
 	return result, nil
 }
 
-func handleError(err error, w http.ResponseWriter) {
+func marshalResponse(resp interface{}, w http.ResponseWriter, status int, r *http.Request, marshalers map[string]ContentMarshaler) error {
+	marshaler, contentType := selectContentMarshaler(r, marshalers)
+	result, err := marshaler.Marshal(resp)
+	if err != nil {
+		return err
+	}
+	writeResult(w, result, status, contentType)
+	return nil
+}
+
+func selectContentMarshaler(r *http.Request, marshalers map[string]ContentMarshaler) (marshaler ContentMarshaler, contentType string) {
+	if accept, found := r.Header["Accept"]; found {
+		var contentTypes []string
+		for ct := range marshalers {
+			contentTypes = append(contentTypes, ct)
+		}
+
+		contentType = goautoneg.Negotiate(accept[0], contentTypes)
```

Comment:
could we use https://github.com/golang/gddo/blob/master/httputil/negotiate.go for negotiation? goautoeng has no updates since 2012 which is pretty old for a golang library imho